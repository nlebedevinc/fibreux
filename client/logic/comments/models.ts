/**
 * This is a definition of domain model for `comments` context.
 *
 * Here we only write classes and types that are shared with the server.
 * In case you need a type that only work on the client,
 * then use a regular typescript type and place it in `/types.ts` file.
 */

import * as ts from 'io-ts'

/**
 * Runtime type, that can be used for schema validation.
 *
 * We call them "models", because they are exchanged between client and server.
 *
 * @see https://github.com/gcanti/io-ts
 * @see https://github.com/aeirola/io-ts-promise
 */
export const RawComment = ts.type({
  'id': ts.number,
  'body': ts.string,
  'email': ts.string,
})

// Static TypeScript type, that can be used as a regular `type`:
export type RawCommentType = ts.TypeOf<typeof RawComment>
