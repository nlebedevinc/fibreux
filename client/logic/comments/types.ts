/**
 * These are inner types, they are only used inside the client.
 *
 * If your exchange some information with a server - make it a model.
 * And place it in `/models.ts` file.
 */

import { RawCommentType } from '~/logic/comments/models'

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number
}

export interface CommentPayloadType {
  commentId: number
  delta: number
}
