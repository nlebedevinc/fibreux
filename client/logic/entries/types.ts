// This is a definition of domain model for `entries` context.

import * as ts from 'io-ts'

// Runtime type, that can be used for schema validation:
export const Entry = ts.type({
  'id': ts.string,
  'description': ts.union([ts.string, ts.null]),
  'person': ts.string,
  'when': ts.union([ts.string, ts.null]),
  'time': ts.union([ts.string, ts.null]),
  'ticket': ts.union([ts.string, ts.null]),
  'project': ts.string,
})

export enum Filters {
  Day = 1,
  Week,
  Month,
}

export const Period = ts.type({
  'from': ts.string,
  'to': ts.string,
})

export type PeriodType = ts.TypeOf<typeof Period>

// Static TypeScript type, that can be used as a regular `type`:
export type EntryType = ts.TypeOf<typeof Entry>

export interface StateType {
  entries: EntryType[]
  activeFilter: number
  activeDate: Date
  selectedEntry: EntryType | null
}
