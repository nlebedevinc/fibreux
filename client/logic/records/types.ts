// This is a definition of domain model for `comments` context.

import * as ts from 'io-ts'

// Runtime type, that can be used for schema validation:
export const RawComment = ts.type({
  'id': ts.number,
  'body': ts.string,
  'email': ts.string,
})

// Static TypeScript type, that can be used as a regular `type`:
export type RawCommentType = ts.TypeOf<typeof RawComment>

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number;
}

export interface CommentPayloadType {
  commentId: number;
  delta: number;
}

export interface StateType {
  comments: CommentType[]
}
