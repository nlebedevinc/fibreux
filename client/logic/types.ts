import { CommentType } from '~/logic/comments/types'

/**
 * We use this state type to enforce the same structure with `vuex-simple`.
 *
 * There are places where we use `Store<StateType>` and it cannot be changed.
 * This is a temporary solution.
 *
 * @see https://github.com/sascha245/vuex-simple/issues/30
 */
export interface RootStateType {
  // Module:
  comments: {
    // Module's state:
    comments: CommentType[]
  }
}
