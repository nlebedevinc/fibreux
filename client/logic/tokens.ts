import { Token } from 'vue-typedi'

/**
 * Here we generate some unique keys to bind our values to.
 *
 * This tokens are unique representation of your dependencies inside the app.
 *
 * @see https://github.com/sascha245/vue-typedi
 * @see https://github.com/typestack/typedi
 */
export default {
  // Comments service:
  'COMMENT_SERVICE': new Token('comment-service'),

  // Axios instance:
  'AXIOS': new Token('axios'),
}
