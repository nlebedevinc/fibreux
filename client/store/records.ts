import { actions } from '~/logic/records/actions'
import { getters } from '~/logic/records/getters'
import { mutations } from '~/logic/records/mutations'
import { state } from '~/logic/records/state'

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
}
