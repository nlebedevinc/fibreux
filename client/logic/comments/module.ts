import { Inject, Injectable } from 'vue-typedi'
import { Action, Mutation, State, Getter } from 'vuex-simple'

import { RawCommentType } from '~/logic/comments/models'
import CommentService from '~/logic/comments/services/api'
import { CommentType, CommentPayloadType } from '~/logic/comments/types'
import tokens from '~/logic/tokens'

@Injectable()
/**
 * Represents a typed Vuex module.
 *
 * @see https://vuex.vuejs.org/guide/modules.html
 * @see https://github.com/sascha245/vuex-simple
 */
export default class CommentsModule {
  // Dependencies

  @Inject(tokens.COMMENT_SERVICE)
  public service!: CommentService

  // State

  @State()
  public comments: CommentType[] = []

  // Getters

  @Getter()
  public get hasComments(): boolean {
    return Boolean(this.comments && this.comments.length > 0)
  }

  // Mutations

  @Mutation()
  public setComments(payload: RawCommentType[]): void {
    const updatedComments: CommentType[] = []

    for (const comment of payload.slice(0, 10)) {
      // We transform RawCommentType in CommentType here:
      updatedComments.push({ ...comment, 'rating': 0 })
    }

    this.comments = updatedComments
  }

  @Mutation()
  public updateRating({ commentId, delta }: CommentPayloadType): void {
    if (!this.comments) return

    const commentIndex = this.comments.findIndex((comment): boolean => {
      return comment.id === commentId
    })

    if (!this.comments || !this.comments[commentIndex]) return

    this.comments[commentIndex].rating += delta
  }

  // Actions

  @Action()
  public async fetchComments(): Promise<RawCommentType[]> {
    const commentsList = await this.service.fetchComments()
    this.setComments(commentsList)
    return commentsList
  }
}
