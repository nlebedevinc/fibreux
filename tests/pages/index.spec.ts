import { mount, createLocalVue } from '@vue/test-utils'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { Container } from 'vue-typedi'
import { Store } from 'vuex'

import { CommentType } from '~/logic/comments/types'
import TypedStore from '~/logic/store'
import tokens from '~/logic/tokens'
import { StateType } from '~/logic/records/types'
import Index from '~/pages/index.vue'

import { storeFactory, commentFactory } from '@/fixtures/vuex'

const localVue = createLocalVue()

describe('unit tests for Index view', () => {
  let store: Store<StateType>
  let comments: CommentType[]

  beforeEach(() => {
    comments = commentFactory.buildList(3)
    store = storeFactory.build(undefined, {
      localVue,
      'state': {
        'comments': { comments },
      },
    })

    // Clearing mocks:
    jest.clearAllMocks()
  })

  test('should have three comments', () => {
    expect.hasAssertions()

    const wrapper = mount(Index, {
      store,
      localVue,
      'propsData': { comments },
    })

    expect(wrapper.findAll('.comment-component')).toHaveLength(3)
  })

  test('should load new comments on actions', async () => {
    expect.hasAssertions()

    const comment = comments[0]
    const wrapper = mount(Index, {
      store,
      localVue,
      'propsData': { comments },
    })
    const typedStore: TypedStore = (wrapper.vm as any).typedStore

    new MockAdapter(axios).onGet('/comments').reply(200, [comment])
    Container.set(tokens.AXIOS, axios)

    await typedStore.comments.fetchComments()

    expect(wrapper.vm.$store.state.comments.comments).toHaveLength(1)
    expect(wrapper.vm.$store.state.comments.comments[0].id)
      .toStrictEqual(comment.id)
    expect(wrapper.findAll('.comment-component')).toHaveLength(1)
  })
})

describe('snapshot test for Index view', () => {
  let store: Store<StateType>
  let comments: CommentType[]

  beforeAll(() => {
    // We need a seed here to be consistent for snapshot testing:
    comments = commentFactory.buildList(3, {}, { 'seed': 1342 })
    store = storeFactory.build(undefined, {
      localVue,
      'state': {
        'comments': { comments },
      },
    })
  })

  test('should match the snapshot', () => {
    expect.hasAssertions()

    const wrapper = mount(Index, {
      store,
      localVue,
    })

    expect(wrapper).toMatchSnapshot()
  })
})
