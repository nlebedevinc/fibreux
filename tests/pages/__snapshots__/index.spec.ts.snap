// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for Index view should match the snapshot 1`] = `
<main class="main">
  <div class="logo">
    <h1><a href="https://github.com/wemake-services/wemake-vue-template">
        wemake-vue-template
      </a></h1>
    <h2><a href="https://wemake-services.gitbook.io/wemake-vue-template/">
        documentation
      </a></h2>
  </div>
  <div class="actions"><button class="reload">
      reload comments
    </button></div>
  <section class="container">
    <div class="comment-component comment-negative">
      <p class="author">
        Alfred.Barton@hotmail.com
      </p>
      <p class="body">
        Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.
      </p>
      <div class="rating"><button class="rating-up">
          +
        </button> <span class="number">-10</span> <button class="rating-down">
          -
        </button></div>
    </div>
    <div class="comment-component comment-negative">
      <p class="author">
        Alfred.Barton@hotmail.com
      </p>
      <p class="body">
        Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.
      </p>
      <div class="rating"><button class="rating-up">
          +
        </button> <span class="number">-10</span> <button class="rating-down">
          -
        </button></div>
    </div>
    <div class="comment-component comment-negative">
      <p class="author">
        Alfred.Barton@hotmail.com
      </p>
      <p class="body">
        Et perspiciatis facilis sed sit labore. Ipsum mollitia sunt occaecati. Expedita ut perspiciatis quod enim autem eum. Unde quibusdam eligendi deleniti consectetur.
      </p>
      <div class="rating"><button class="rating-up">
          +
        </button> <span class="number">-10</span> <button class="rating-down">
          -
        </button></div>
    </div>
  </section>
</main>
`;
